### PR TITLE
aligh with rcl that a rosout publisher of a node might not exist

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
align with https://github.com/ros2/rcl/pull/1115

### CODE REVIEW

File: rclpy/src/rclpy/_rclpy_logging.cpp [hunk 1]
```diff
@@ -206,14 +206,18 @@ rclpy_logging_get_logging_directory()
...
-  if (RCL_RET_OK != rcl_ret) {
-    rcutils_reset_error();
-    throw std::runtime_error("Failed to call rcl_logging_rosout_add_sublogger");
+  if (RCL_RET_OK == rcl_ret) {
+    return true;
+  } else if (RCL_RET_NOT_FOUND == rcl_ret) {
+    rcl_reset_error();
+    return false;
+  } else {
+    throw rclpy::append_rcl_error("Failed to call rcl_logging_rosout_add_sublogger");
```
Comment (fujitatomoya) [inline rclpy/src/rclpy/_rclpy_logging.cpp:220]: having exception and false looks redundant? i think for user perspective, it is not really clear the difference between having exception and false via this call.
how about keeping this `void` and if `RCL_RET_NOT_FOUND` print warning (there will no rosout) w/o exception, and everything else goes to exception?

File: rclpy/test/test_rosout_subscription.py [hunk 2]
```diff
@@ -114,13 +114,6 @@ def call_logger(logger):
         self.executor.spin_until_future_complete(self.fut, 3)
         self.assertTrue(self.fut.done())
 
-    def test_node_logger_not_exist(self):
-        node = rclpy.create_node('test_extra_node', context=self.context)
-        logger = node.get_logger()
-        node = None
-        with self.assertRaises(RuntimeError):
-            logger.get_child('child')
```
Comment (fujitatomoya) [inline rclpy/test/test_rosout_subscription.py:122]: then this test needs to make sure there will be no exception but warning?

File: rclpy/src/rclpy/_rclpy_logging.cpp (hunk 1 above)
Comment (iuhilnehc-ynos) [reply rclpy/src/rclpy/_rclpy_logging.cpp:220]: Using a boolean is to make sure https://github.com/ros2/rclpy/blob/059bac65707c0e605a08bcf9451738f891ed0d26/rclpy/rclpy/impl/rcutils_logger.py#L238-L239 not to update `logger.logger_sublogger_namepair` if it's false(there is no parent logger existent),
and then it's unnecessary to call `_rclpy.rclpy_logging_rosout_remove_sublogger` https://github.com/ros2/rclpy/blob/059bac65707c0e605a08bcf9451738f891ed0d26/rclpy/rclpy/impl/rcutils_logger.py#L222-L223

If there exist other errors (e.g. `RCUTILS_RET_INVALID_ARGUMENT`, etc) when calling rcutils_hash_map_get, `rclpy_logging_rosout_add_sublogger` just throw the exceptions similar to other APIs in current file.

If only printing warning for `RCL_RET_NOT_FOUND`, there exists some unnecessary actions(`add logger.logger_sublogger_namepair` and `_rclpy.rclpy_logging_rosout_remove_sublogger`).

File: rclpy/test/test_rosout_subscription.py [hunk 3]
```diff
@@ -114,13 +114,6 @@ def call_logger(logger):
         self.executor.spin_until_future_complete(self.fut, 3)
         self.assertTrue(self.fut.done())
 
-    def test_node_logger_not_exist(self):
```
Comment (clalancette) [inline rclpy/test/test_rosout_subscription.py:117]: I think we want to add a test to ensure that the situation from https://github.com/ros2/rclpy/issues/1195 doesn't occur again.  The following code (which needs to be adapted into a test) reproduced for me:

```
import rclpy
from rclpy.action import ActionServer

from example_interfaces.action import Fibonacci

class MyClass:
    def __init__(self, node):
        self.action_server = ActionServer(
            node,
            Fibonacci,
            "/fibonacci_action_1",
            self.execute_callback,
        )

    def execute_callback(self):
        print("Execute callback")

rclpy.init()
node = rclpy.create_node("mynode", enable_rosout=False)
x = MyClass(node)
node.destroy_node()
rclpy.try_shutdown()
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I think we need to add a test for this situation, but it otherwise looks good.

File: rclpy/test/test_rosout_subscription.py (hunk 3 above)
Comment (iuhilnehc-ynos) [reply rclpy/test/test_rosout_subscription.py:117]: Thanks. 
I used the similar logic to test the case. Updated in https://github.com/ros2/rclpy/pull/1196/commits/fc84c5e18c1655d2a0ed88dc4a04ef3d295d8459.

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI for this is in https://github.com/ros2/rcl/pull/1115#issuecomment-1809415104

**Pull request merged**